File: backend/app/api/schemas/context_status.py
from __future__ import annotations

from typing import Literal, Optional

from pydantic import BaseModel


class ContextStatusUsageResponse(BaseModel):
    input_tokens: Optional[int] = None
    output_tokens: Optional[int] = None
    total_tokens: Optional[int] = None
    cached_input_tokens: Optional[int] = None
    cached_output_tokens: Optional[int] = None
    reasoning_tokens: Optional[int] = None


class ContextStatusResponse(BaseModel):
    model_id: Optional[str] = None
    max_tokens: Optional[int] = None
    max_tokens_source: Optional[str] = None
    reserved_output_tokens: Optional[int] = None
    estimated_input_tokens: Optional[int] = None
    estimated_total_tokens: Optional[int] = None
    estimated_remaining_tokens: Optional[int] = None
    estimated_usage_ratio: Optional[float] = None
    near_limit: bool = False
    compaction_recommended: bool = False
    source: Literal["estimated_pre_run", "estimated_in_flight", "estimated_plus_actual"] = "estimated_pre_run"
    actual_usage: Optional[ContextStatusUsageResponse] = None

File: frontend-reshet/src/services/context-status.ts
export interface ContextStatusUsage {
  input_tokens?: number | null;
  output_tokens?: number | null;
  total_tokens?: number | null;
  cached_input_tokens?: number | null;
  cached_output_tokens?: number | null;
  reasoning_tokens?: number | null;
}

export interface ContextStatus {
  model_id?: string | null;
  max_tokens?: number | null;
  max_tokens_source?: string | null;
  reserved_output_tokens?: number | null;
  estimated_input_tokens?: number | null;
  estimated_total_tokens?: number | null;
  estimated_remaining_tokens?: number | null;
  estimated_usage_ratio?: number | null;
  near_limit: boolean;
  compaction_recommended: boolean;
  source: "estimated_pre_run" | "estimated_in_flight" | "estimated_plus_actual";
  actual_usage?: ContextStatusUsage | null;
}

const SOURCE_PRIORITY: Record<ContextStatus["source"], number> = {
  estimated_pre_run: 1,
  estimated_in_flight: 2,
  estimated_plus_actual: 3,
};

function numericOrNull(value: number | null | undefined): number | null {
  return typeof value === "number" && Number.isFinite(value) ? value : null;
}

function statusWeight(status: ContextStatus | null | undefined): [number, number, number] {
  if (!status) {
    return [0, 0, 0];
  }
  return [
    SOURCE_PRIORITY[status.source] || 0,
    numericOrNull(status.actual_usage?.total_tokens) || 0,
    numericOrNull(status.estimated_total_tokens) || 0,
  ];
}

export function mergeContextStatus(
  current: ContextStatus | null | undefined,
  incoming: ContextStatus | null | undefined,
): ContextStatus | null {
  const normalizedCurrent = current || null;
  const normalizedIncoming = incoming || null;
  if (!normalizedIncoming) {
    return normalizedCurrent;
  }
  if (!normalizedCurrent) {
    return normalizedIncoming;
  }

  const currentWeight = statusWeight(normalizedCurrent);
  const incomingWeight = statusWeight(normalizedIncoming);
  if (
    incomingWeight[0] > currentWeight[0]
    || (incomingWeight[0] === currentWeight[0] && incomingWeight[1] > currentWeight[1])
    || (
      incomingWeight[0] === currentWeight[0]
      && incomingWeight[1] === currentWeight[1]
      && incomingWeight[2] >= currentWeight[2]
    )
  ) {
    return normalizedIncoming;
  }
  return normalizedCurrent;
}
